src/assign.rs

```diff
@@ -58,25 +65,176 @@ pub struct Assignment<'a> {
     ///
     /// ## Example
     /// ```rust
-    /// use serde_json::json;
-    /// use jsonptr::{Pointer, Assign};
+    /// # use serde_json::json;
+    /// # use jsonptr::{Pointer, Assign};
     /// let mut data = json!({ "foo": ["zero"] });
     /// let mut ptr = Pointer::from_static("/foo/-");
     /// let assignment = data.assign(&mut ptr, "one").unwrap();
     /// assert_eq!(assignment.assigned_to, Pointer::from_static("/foo/1"));
     /// ```
-    pub assigned_to: PointerBuf,
+    pub assigned_to: &'p Pointer,
+
+    /// The value that was replaced, if any.
+    pub replaced: Option<Value>,
+}
+
+pub(crate) fn assign_value<'p, 'd>(
+    mut remaining: &'p Pointer,
+    mut dest: &'d mut Value,
+    mut src: Value,
+) -> Result<Assignment<'p, 'd>, AssignError> {
+    let mut offset = 0;
+    let full_ptr = remaining;
+    let mut replaced;
+    let mut returned_src;
+    while let Some((token, tail)) = remaining.split_front() {
+        offset = offset + 1 + token.encoded().len();
+        let (assigned_to, _) = full_ptr
+            .split_at(offset)
+            .unwrap_or((Pointer::root(), Pointer::root()));
+        Assigned {
+            dest,
+            remaining,
+            replaced,
+            src: returned_src,
+        } = match dest {
+            Value::Array(array) => assign_to_array(token, tail, array, src, offset)?,
+            Value::Object(obj) => assign_to_object(token, tail, obj, src)?,
+            _ => assign_to_scalar(remaining, dest, src)?,
+        };
+        if remaining.is_root() {
+            return Ok(Assignment {
+                assigned: dest,
+                assigned_to,
+                replaced,
+            });
+        }
+        src = returned_src.unwrap();
+    }
+    // Pointer is root, we can replace `dest` directly
+    let replaced = Some(core::mem::replace(dest, src.into()));
+    Ok(Assignment {
+        assigned: dest,
+        replaced,
+        assigned_to: remaining,
+    })
+}
+fn create_value_path(mut path: &Pointer, mut src: Value) -> Result<Value, AssignError> {
+    while let Some((token, tail)) = path.split_front() {
+        path = tail;
+        match path.as_str() {
+            "0" | "-" => {
+                src = Value::Array(vec![src]);
+            }
+            _ => {
+                let mut obj = Map::new();
+                obj.insert(token.to_string(), src);
+                src = Value::Object(obj);
+            }
+        }
+    }
+    Ok(src)
+}
+
+struct Assigned<'p, 'd> {
+    remaining: &'p Pointer,
+    dest: &'d mut Value,
+    replaced: Option<Value>,
+    src: Option<Value>,
+}
+
+fn assign_to_array<'p, 'd>(
+    token: Token<'_>,
+    remaining: &'p Pointer,
+    dest: &'d mut Vec<Value>,
+    src: Value,
+    offset: usize,
+) -> Result<Assigned<'p, 'd>, AssignError> {
+    let idx = token
+        .to_index()
+        .with_context(|_| assign_error::FailedToParseIndexSnafu { offset })?
```

Comment:
Feels wrong to ignore the original error (here and elsewhere); I think SNAFU supports error sources?

Reply:
It does and it is. snafu's context system is a bit weird at first but that is wrapping the original error.

The `Snafu` proc macro generates a `Snafu` struct for each variant of an enum error or a single struct for a struct error. The fields on it are anything explicit (backtrace is implicit + [you can create your own](https://docs.rs/snafu/latest/snafu/derive.Snafu.html#controlling-implicitly-generated-data)). Then `context` or `with_context` to inject the source error and implicit data.

shepmaster has written a really good guide here: https://docs.rs/snafu/latest/snafu/guide/index.html

Reply:
One last thing on snafu, its [reporting](https://docs.rs/snafu/latest/snafu/attr.report.html) is incredibly nice. It's more for a bin but using snafu at the lib level enables it.

I'm not certain how critical it is to use snafu at the lib level for reporting at the bin level but it's a nice bonus feature.

Reply:
> One last thing on snafu, its [reporting](https://docs.rs/snafu/latest/snafu/attr.report.html) is incredibly nice.

That's one thing I actually don't like about `snafu`, it tries to be too many things at the same time. It's a utility for deriving error implementations like `thiserror`, but also bundles its own reporting framework like `anyhow` or `eyre`. I generally think it makes more sense to leave reporting to applications, and have libraries only concern themselves with producing good errors. This separation is a great design decision in the `tracing` crate IMO.

> It does and it is. snafu's context system is a bit weird at first but that is wrapping the original error.

I actually don't like this either 😅 . I don't like implicit magic stuff that gets injected behind the scenes. Generally I believe the intent of code should be understandable without having to look up documentation.

> The Snafu proc macro generates a ___Snafu (where ___ is the name of the error, sans "Error" if it has the suffix) struct for each variant of an enum error or a single struct for a struct error.

This is one I probably like least of all. This naming convention is terrible IMO. In a few cases a suffix like `Error` makes sense to qualify an error type name, when otherwise it would be ambiguous (e.g. `ParseError`), but Snafu doesn't mean anything by itself, and getting it injected unconditionally makes it likely someone will make a compromise, choosing an inferior name for their error just so it doesn't clash with the suffix.

But to be clear, this is all just my personal opinion. I'm sharing as a way to provide a different perspective, but don't consider it part of the code review. I'd never reject a PR based on personal opinion (unless it violated some prior convention).

Reply:
> That's one thing I actually don't like about `snafu`, it tries to be too many things at the same time. It's a utility for deriving error implementations like `thiserror`, but also bundles its own reporting framework like anyhow or eyre.

That's fair. In `snafu`'s defense, the formatting and reporting is only generated in the event you utilize the macro on your main. The formatting itself is fairly minimal simply relies on the appropriate chain of source errors being available to unwrap, for lack of better word.

It does do a lot though, I guess. It sort of merges `thiserror` and `anyhow` through `Whatever`. The idea being, I believe, that both strategies are present to allow folks to gradually step up to proper errors from opaque.

> This is one I probably like least of all. This naming convention is terrible IMO. 

I'll be honest, I'm not a huge fan of it either. If I were to speculate, I'd say he went with the naming convention to ensure he had the least chance of collision on existing artifacts within the source.

It is configurable though via the [`context` attribute](https://docs.rs/snafu/latest/snafu/derive.Snafu.html#controlling-context) for the derive.

> But to be clear, this is all just my personal opinion. I'm sharing as a way to provide a different perspective, but don't consider it part of the code review. I'd never reject a PR based on personal opinion (unless it violated some prior convention).

Oh, no, I know. To be clear, I greatly appreciate your opinions. I very much appreciate your questioning of this. I had (and have) doubts about the added weight & complexity before you voiced your concerns.

Reply:
> I actually don't like this either 😅 . I don't like implicit magic stuff that gets injected behind the scenes. Generally I believe the intent of code should be understandable without having to look up documentation.

I completely get where you're coming from. I'm generally avidly opposed to magic. For example, a lot of people _love_ rocket but it is far too 🪄 ✨ for me. At first glance, `snafu` seemed the same. It took me using it to finally come around and appreciate the ergonomics of it.

Having said all that, I think I'll remove it and go back to hand-rolled errors. This is a small, primitive crate that probably shouldn't burden the ecosystem with unnecessary dependencies. While `Backtrace` is nice, its probably more of a bin concern or, at min, a consuming lib.

Reply:
Yeah, didn't want to push hard against it, but my gut feeling is that unless one is building a complex crate that already has a ton of dependencies, staying lean is best. It really doesn't matter most of the time for consumers, but when it does it's a lot of pain. I'm less concerned about the choice of `snafu` vs `thiserror` than about that.